Feature/queue rework

PanopticaRising opened the pull request.

PanopticaRising committed:
fix: Release topic lock when all processing is done.

Added more configurations and logs.

PanopticaRising committed:
feat: Added priorities support based on configuration values.

refactor: Changed some import syntax, added some fixes for logging from the page, moved globals into their own file, same for interfaces on routes.

PanopticaRising committed:
feature: Security fix for pug.

PanopticaRising committed:
fix: Bug where the Zip wasn't waiting for the lock before printing things, and where the conditional for the lock creation was incorrect.

PanopticaRising committed:
fix: Added comparator to HeapHelper to fix heapify.

Updated the aync-mutex to point to our own version instead of the upstream.

PanopticaRising committed:
fix: Updated the Github URL for our version of async-mutex.

Updated logs to be more helpful with tracking TopicIds.

red-tt-bear commented on package-lock.json:
```diff
@@ -837,14 +837,14 @@
-      "version": "7.12.11",
-      "resolved": "https://registry.npmjs.org/@babel/parser/-/parser-7.12.11.tgz",
-      "integrity": "sha512-N3UxG+uuF4CMYoNj8AhnbAcJF0PiuJ9KHuy1lQmkYsxTer/MAH9UBNHsBoAX/4s6NvlDD047No8mYVGGzLL4hg=="
+      "version": "7.13.9",
```
Did you accidentally commit the package-lock.json after it was regenerated? There are a lot of changes here

red-tt-bear commented on src/routes/index.ts:
```diff
@@ -1,79 +1,60 @@
+    // If this is one of the first N PDFs, give elevated priority.
+    if (cheatingInMemoryStorage[topic].pdfPromises.length < configurations.app.highPriorityTabsPerTopic) {
+        logger.debug(`[${topic}] Assigning an initial high priority of 99 to ${body.firstName}`);
+        newPriority.prio = 99;
```
I thought we were going to change this to Number.Max?

red-tt-bear commented on src/routes/index.ts:
```diff
@@ -99,11 +80,20 @@ router.get('/', async (_req, _res, next) => {
         await Promise.allSettled(cheatingInMemoryStorage[topicId].pdfPromises);
     } catch (e) {
         logger.error('Zip was requested before any PDFs were!', e);
+        throw new Error('Zip called too soon.');
```
Throwing an error with the same message in two different places will cause confusion

tommy-lettieri reviewed (changes requested):
package lock should be fine since we are going to need thorough testing anyway
the duplicate error messages should be changed

PanopticaRising commented on src/routes/index.ts:
```diff
@@ -1,79 +1,60 @@
+    const newPriority: PDFPriorityData = { prio: 0, topicId: topic, profUUID: body.professorUUID, firstName: body.firstName };
+
+    // If this is one of the first N PDFs, give elevated priority.
+    if (cheatingInMemoryStorage[topic].pdfPromises.length < configurations.app.highPriorityTabsPerTopic) {
```
Items are pushed to this array on line 47 below this, so there will always be a time when this is checked and there are no items in the array.

red-tt-bear commented on src/routes/index.ts:
```diff
@@ -1,79 +1,60 @@
 
-    addPDFToZip(cheatingInMemoryStorage[topic].zipObject.archive, pdfPromise, topic)
-    .catch(e => logger.error('Route failed to add pdf to zip', e));
+    await cheatingInMemoryStorage[topic].lock;
```
What happens if an error occurs after the request responds

i.e. lock times outs

PanopticaRising committed:
fix: Added more postBack Error conditions.

Added timeouts for concurrentTopics and error handling if that's ever hit.

Changed the initial priority to 999999 so it's high enough for most use cases but easy enough to read/debug.
